pkg/controller/workspacerouting/solvers/openshift_oauth_solver.go

```diff
@@ -0,0 +1,189 @@
+	proxyServices := getServicesForEndpoints(proxyPorts, workspaceMeta)
+	for idx := range proxyServices {
+		proxyServices[idx].Annotations = map[string]string{
+			"service.alpha.openshift.io/serving-cert-secret-name": "proxy-tls",
+		}
+	}
+	discoverableServices := getDiscoverableServicesForEndpoints(proxyPorts, workspaceMeta)
+	services := append(proxyServices, discoverableServices...)
+
+	routes, proxyEndpoints, podAdditions := s.getProxyRoutes(proxy, workspaceMeta, portMappings)
+	for machineName, machineEndpoints := range proxyEndpoints {
+		exposedEndpoints[machineName] = append(exposedEndpoints[machineName], machineEndpoints...)
+	}
+
+	return RoutingObjects{
+		Services:         services,
+		Ingresses:        defaultIngresses,
+		Routes:           routes,
+		PodAdditions:     podAdditions,
+		ExposedEndpoints: exposedEndpoints,
+	}
+}
+
+func (s *OpenShiftOAuthSolver) getProxyRoutes(
+	endpoints map[string][]v1alpha1.Endpoint,
+	workspaceMeta WorkspaceMetadata,
+	portMappings map[string]proxyEndpoint) ([]routeV1.Route, map[string][]v1alpha1.ExposedEndpoint, *v1alpha1.PodAdditions) {
+
+	var routes []routeV1.Route
+	var exposedEndpoints = map[string][]v1alpha1.ExposedEndpoint{}
+	var podAdditions *v1alpha1.PodAdditions
+
+	for machineName, machineEndpoints := range endpoints {
+		for _, upstreamEndpoint := range machineEndpoints {
+			proxyEndpoint := portMappings[upstreamEndpoint.Name]
+			endpoint := proxyEndpoint.publicEndpoint
+			targetEndpoint := intstr.FromInt(int(endpoint.Port))
+			endpointName := common.EndpointName(endpoint.Name)
+			hostname := common.EndpointHostname(workspaceMeta.WorkspaceId, endpointName, endpoint.Port, workspaceMeta.IngressGlobalDomain)
+			url := fmt.Sprintf("https://%s", hostname)
+
+			// NOTE: openshift oauth-proxy only supports listening on a single port; as a result, we can't proxy more than
+			// one endpoint or we run into cookie issues (each proxied port gets a container and sets a cookie in the browser
+			// once auth is completed).
+			var tls *routeV1.TLSConfig = nil
+			if endpoint.Attributes[v1alpha1.SECURE_ENDPOINT_ATTRIBUTE] == "true" {
+				if endpoint.Attributes[v1alpha1.TYPE_ENDPOINT_ATTRIBUTE] == "terminal" {
+					tls = &routeV1.TLSConfig{
+						Termination:                   routeV1.TLSTerminationEdge,
+						InsecureEdgeTerminationPolicy: routeV1.InsecureEdgeTerminationPolicyRedirect,
+					}
+				} else {
+					tls = &routeV1.TLSConfig{
+						Termination:                   routeV1.TLSTerminationReencrypt,
+						InsecureEdgeTerminationPolicy: routeV1.InsecureEdgeTerminationPolicyRedirect,
+					}
+				}
+			}
+			routes = append(routes, routeV1.Route{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      common.RouteName(workspaceMeta.WorkspaceId, endpointName),
+					Namespace: workspaceMeta.Namespace,
+					Labels: map[string]string{
+						config.WorkspaceIDLabel: workspaceMeta.WorkspaceId,
+					},
+				},
+				Spec: routeV1.RouteSpec{
+					Host: hostname,
```

Comment:
I wonder if we should specify hostname at all? I know it helps to solve issues with route lengh and uniqueness but I don't think it's possible on the secure openshift instances, like OSIO
![Screenshot_20200403_155155](https://user-images.githubusercontent.com/5887312/78362611-159e7480-75c3-11ea-8cc2-0415401942c1.png)
Or you think that controller must be able to set host?

Reply:
The main issue is that hostname must be set on ingresses; it's done here mostly because the route code is a port of the ingress code; we *should* drop it here, but the PR will need to be reworked to create only routes and not ingresses when running on OpenShift.

Also, I don't have a minishift instance handy for testing, but I wonder how it would work there since we currently rely on using `$(minishift ip).nip.io`.